src/bootstrap/native.rs

```diff
@@ -803,7 +803,21 @@ impl Step for Lld {
         }
         let target = self.target;
 
-        let llvm_config = builder.ensure(Llvm { target: self.target });
+        let llvm_config = builder.ensure(Llvm { target });
+
+        // The `dist` step packages LLD next to LLVM's binaries for download-ci-llvm. The root path
+        // we usually expect here is `./build/$triple/ci-llvm/`, with the binaries in its `bin`
+        // subfolder. We check if that's the case, and if LLD's binary already exists there next to
+        // `llvm-config`: if so, we can use it instead of building LLVM/LLD from source.
+        let ci_llvm_bin = llvm_config.parent().unwrap();
+        if ci_llvm_bin.is_dir() && ci_llvm_bin.file_name().unwrap() == "bin" {
+            let lld_path = ci_llvm_bin.join(exe("lld", target));
+            if lld_path.exists() {
+                // The following steps copying `lld` as `rust-lld` to the sysroot, expect it in the
+                // `bin` subfolder of this step's out dir.
+                return ci_llvm_bin.parent().unwrap().to_path_buf();
```

Comment:
Why does LLD need to be special cased this way? It looks like the `Llvm` step just builds unconditionally. Should we put this logic in the `dist` step instead of here?

Reply:
This logic already is in the `dist` step for the [`RustDev` component](https://github.com/rust-lang/rust/blob/8de4b138455add55bde6de5553a933a2ab79b71f/src/bootstrap/dist.rs#L2081-L2085).  `dist` will optionally package `lld` in the `RustDev` component if it can find it on disk, but the `native::Lld` step is always expecting to build it from source. 

However, the `native::Llvm` step doesn't [just build unconditionally](https://github.com/rust-lang/rust/blob/8de4b138455add55bde6de5553a933a2ab79b71f/src/bootstrap/native.rs#L255-L259), it [can work with `download-ci-llvm`](https://github.com/rust-lang/rust/blob/8de4b138455add55bde6de5553a933a2ab79b71f/src/bootstrap/native.rs#L68-L77) and this change is the dual to those, making `native::Lld` able to read what `dist` already writes to the component.

Reply:
> However, the native::Llvm step doesn't [just build unconditionally](https://github.com/rust-lang/rust/blob/8de4b138455add55bde6de5553a933a2ab79b71f/src/bootstrap/native.rs#L255-L259), it [can work with download-ci-llvm](https://github.com/rust-lang/rust/blob/8de4b138455add55bde6de5553a933a2ab79b71f/src/bootstrap/native.rs#L68-L77) and this change is the dual to those, making native::Lld able to read what dist already writes to the component.

:+1: makes sense, thanks for explaining :)